packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -944,26 +924,58 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 	}
 
 	public async getPendingBlobs(waitBlobsToAttach?: boolean): Promise<IPendingBlobs | undefined> {
-		if (waitBlobsToAttach) {
-			await this.stashPendingBlobs();
-		}
-
-		if (this.pendingBlobs.size === 0) {
-			return;
-		}
-
-		const blobs = {};
-		for (const [key, entry] of this.pendingBlobs) {
-			blobs[key] = {
-				blob: bufferToString(entry.blob, "base64"),
-				storageId: entry.storageId,
-				attached: entry.attached,
-				acked: entry.acked,
-				minTTLInSeconds: entry.minTTLInSeconds,
-				uploadTime: entry.uploadTime,
-			};
-		}
-		return blobs;
+		return PerformanceEvent.timedExecAsync(
+			this.mc.logger,
+			{ eventName: "GetPendingBlobs" },
+			async () => {
+				if (this.pendingBlobs.size === 0) {
+					return;
+				}
+				const blobs = {};
+				const localBlobs = new Set<PendingBlob>();
+				while (localBlobs.size < this.pendingBlobs.size) {
+					const attachBlobsP: Promise<void>[] = [];
+					for (const [id, entry] of this.pendingBlobs) {
+						if (!localBlobs.has(entry)) {
+							localBlobs.add(entry);
+							if (waitBlobsToAttach) {
+								if (!entry.opsent) {
+									this.sendBlobAttachOp(id, entry.storageId);
+								}
+								entry.handleP.resolve(this.getBlobHandle(id));
+								if (!entry.attached) {
```

Comment:
lets move this in the promise, as it shouldn't be true here, since we just returned the handled, but it could change between promise create, and promise run. we want to check it before we call `this.on("blobAttached", onBlobAttached);` and just resolve if it happen to be true, only if false should we hook up the event handler